### PR TITLE
Java.Lang.Double

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: java/lang/Double/Double.cpp [hunk 1]
```diff
@@ -74,7 +74,8 @@ Double Double::parseDouble(String target) {
  * @return String
  */
 string Double::toString() const {
-	return string_from_double(this->original);
+	string resultToString = string_from_double(this->original);
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: You should implement this toString method the same as Integer

File: java/lang/Double/Double.cpp [hunk 2]
```diff
@@ -92,7 +96,8 @@ char Double::charValue() const {
  * @return CString
  */
 string Double::stringValue() const {
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: You can call toString to make thing easier

File: java/lang/Double/Double.cpp [hunk 3]
```diff
@@ -101,7 +106,7 @@ string Double::stringValue() const {
  * @return short
  */
 short Double::shortValue() const {
-	return this->original;
+    return this->original;
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: You should cast to **(short)** to make sense

File: java/lang/Double/Double.cpp [hunk 4]
```diff
@@ -146,7 +151,7 @@ double Double::doubleValue() const {
  * @return Double
  */
 Double Double::operator+(const Double &target) {
-	return this->original + target.original;
+    return this->original + target.original;
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: Cast to **Double(...)** to make sense

**Review by dthongvl — CHANGES_REQUESTED** [review]
Please check and implement all methods in
https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html

File: java/lang/Double/Double.cpp (hunk 4 above)
Comment (tiny-express) [reply java/lang/Double/Double.cpp]: How about this one ?

File: java/lang/Double/Double.cpp [hunk 5]
```diff
@@ -208,57 +226,265 @@ Double Double::operator/(const Double &target) {
...
-/**
- * Compare this Double is more than target
- *
- * @return bool
- */
 boolean Double::operator>(const Double &target) const {
-	return this->original > target.original;
+//    if (isNaN(target.doubleValue()) || isInfinite(target.doubleValue()))
+//        return -1;
+    return (boolean ) (this->original > target.original);
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: Remove space **(boolean)**

File: java/lang/Double/Double.cpp [hunk 6]
```diff
@@ -208,57 +226,265 @@ Double Double::operator/(const Double &target) {
...
+    return (boolean) (this->original && target.original);
+}
+
+boolean Double::operator||(const Double &target) const {
+//    if (isNaN(target.doubleValue()) || isInfinite(target.doubleValue()))
+//        return -1;
+    return (boolean) (this->original || target.original);
+}
+
+/// ---------- Assignment Operators ----------
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: Remove all comment ///

File: java/lang/Double/Double.cpp [hunk 7]
```diff
@@ -208,57 +226,265 @@ Double Double::operator/(const Double &target) {
...
+}
+
+long Double::doubleToLongBits(double valueDouble) {
+    if (isNaN(valueDouble) || isInfinite(valueDouble))
+        return -9999;
+    long doubleToLongBitsResult = doubleToRawLongBits(valueDouble);
+
+    // Check for NaN based on values of bit fields, maximum
+    // exponent and nonzero significand.
+    long condition1 = doubleToLongBitsResult & EXP_BIT_MASK;
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp]: Please rename var to make sense

File: java/lang/Double/Double.cpp [hunk 8]
```diff
@@ -208,57 +226,265 @@ Double Double::operator/(const Double &target) {
...
+    unsigned long rightShiftBits = (unsigned long)bits >> 32;
+    return  (bits ^ rightShiftBits);
+}
+
+boolean Double::isFinite(double valueDouble) {
+    return (Math::abs(valueDouble) <= MAX_VALUE_DOUBLE);
+}
+
+boolean Double::isInfinite(double valueDouble) {
+    boolean condition1 = (valueDouble == POSITIVE_INFINITY_DOUBLE);
```
Comment (dthongvl) [inline java/lang/Double/Double.cpp:405]: Please rename var to make sense

File: java/lang/Double/Double.hpp [hunk 9]
```diff
@@ -31,47 +31,855 @@
...
+
+//            /**
+//             * Decrement operator,
+//             * decreases integer value by one
+//             *
+//             * @return Double
+//             */
+//            Double operator--(const Double &target) const;
+
+            /// ---------- Relational Operators ----------
```
Comment (dthongvl) [inline java/lang/Double/Double.hpp]: Please remove any comment ///

File: java/lang/Double/Double.hpp [hunk 10]
```diff
@@ -31,47 +31,855 @@
...
+
+            /**
+             * Double to String
+             *
+             * @return String
+             */
+            string toString() const;
+
+            /**
+             * Returns a string representation of the {@code double}
```
Comment (dthongvl) [inline java/lang/Double/Double.hpp:405]: Please shorten this, just comment useful information and remove weird character

File: java/lang/Double/DoubleTest.cpp [hunk 11]
```diff
@@ -31,79 +31,1215 @@ extern "C" {
...
+    ASSERT_DBL_NEAR(expectedResultOperator.doubleValue(), actualResultOperator.doubleValue());
+
+    // Test Operator *
+    variableArithmeticOperator1 =  2.0;
+    variableArithmeticOperator2 =  13.02;
+    expectedResultOperator = 26.04;
+    actualResultOperator = variableArithmeticOperator1 * variableArithmeticOperator2;
+    ASSERT_DBL_NEAR(expectedResultOperator.doubleValue(), actualResultOperator.doubleValue());
+
+    // Test Operator /
```
Comment (dthongvl) [inline java/lang/Double/DoubleTest.cpp:103]: What will happen if divide by zero?

File: java/lang/Double/DoubleTest.cpp [hunk 12]
```diff
@@ -31,79 +31,1215 @@ extern "C" {
...
+    expectedResultOperator = 13.02;
+    actualResultOperator = (variableAssignmentOperator1 /= variableAssignmentOperator2);
+    ASSERT_DBL_NEAR(expectedResultOperator.doubleValue(), actualResultOperator.doubleValue());
 }
 
 TEST (JavaLang, DoubleParseDouble) {
-	// Given value for Double constructor and assign value - Return double
-	Double validDoublePointer = Double::parseDouble("6");
-	ASSERT_EQUAL(6, validDoublePointer.doubleValue());
+#ifdef __APPLE__
```
Comment (dthongvl) [inline java/lang/Double/DoubleTest.cpp]: Why return here in OSX?

**Review by dthongvl — CHANGES_REQUESTED** [review]
Recommend you should move test XXXValue() to the end of file because it nearly not wrong.
It will help reviewer.

**Comment by dthongvl** [comment]
Approved at this time.
Closing to wait for other class to complete

**Review by dthongvl — APPROVED** [review]

**Pull request merged**